### PR TITLE
add deprecation to `NAME_KEY`

### PR DESCRIPTION
**Pull request by bekzod**
since `NAME_KEY` is still widely used https://emberobserver.com/code-search?codeQuery=NAME_KEY

### CODE REVIEW

File: packages/ember-runtime/lib/system/core_object.js [hunk 1]
```diff
@@ -608,7 +609,7 @@ CoreObject.PrototypeMixin = Mixin.create({
     let hasToStringExtension = typeof this.toStringExtension === 'function';
     let extension = hasToStringExtension ? `:${this.toStringExtension()}` : '';
 
-    let ret = `<${getName(this) || FACTORY_FOR.get(this) || this.constructor.toString()}:${guidFor(
+    let ret = `<${ this[NAME_KEY] || getName(this) || FACTORY_FOR.get(this) || this.constructor.toString()}:${guidFor(
```
Comment (bekzod) [inline packages/ember-runtime/lib/system/core_object.js]: reverted old behavior here, as some addon depend on it right now

**Comment by bekzod** [comment]
@krisselden good idea ?

**Comment by rwjblue** [comment]
IMHO, nearly all of these cases would be better implemented via a `.toString` on the class/mixin/etc (which would be backwards compatible). Also, most of these addons use it speculatively, and we ensured they didn't error by continuing to _have_ an `Ember.NAME_KEY` present.  

tldr; unless we are aware of actual breakage, I'd rather not bring back the lookup for `this[NAME_KEY]`...

**Comment by bekzod** [comment]
Got it, should we at least keep deprecation ?

**Comment by rwjblue** [comment]
Yes, that seems good to me, though it will need to tell folks to use `.toString` instead.

**Comment by krisselden** [comment]
The usage of this.NAME_KEY in CoreObject instance method toString is likely never used, and the examples linked to are trying to add it to the constructor for classToString behavior.  I don't remember why we did the NAME_KEY in there, I think it may have been how the factory at one point set name by passing it to create().  I'd like to redo the string stuff around just constructor.name and factory name, remove the namespace stuff.

**Comment by krisselden** [comment]
I see, you want to propose setName be used instead of this pattern, we use constructor to string so if they implement a toString method on the constructor, it will be used instead.

**Comment by bekzod** [comment]
anyway public usage should be deprecated right ? if not then you can close it

File: packages/ember/index.js [hunk 2]
```diff
@@ -201,7 +201,20 @@ Ember.canInvoke = utils.canInvoke;
...
 Ember.uuid = utils.uuid;
-Ember.NAME_KEY = utils.NAME_KEY;
+Ember.setName = utils.setName;
+
+Object.defineProperty(Ember, 'NAME_KEY', {
+  enumerable: false,
+  get() {
+    deprecate('Using `Ember.NAME_KEY` is deprecated, override `.toString` instead', false, {
+      id: 'ember-name-key-usage',
+      until: '4.0.0',
```
Comment (rwjblue) [inline packages/ember/index.js]: Lets say 3.5.0 here, I don't think `NAME_KEY` was ever public API...

**Review by rwjblue — CHANGES_REQUESTED** [review]
One minor tweak, but seems good to me after that...

File: packages/ember/index.js (hunk 2 above)
Comment (bekzod) [reply packages/ember/index.js]: done

**Review by rwjblue — APPROVED** [review]
Looks good, would you mind doing a deprecation-app entry for this?

File: packages/ember/tests/reexports_test.js [hunk 3]
```diff
@@ -34,6 +34,12 @@ moduleFor(
       }, /EXTEND_PROTOTYPES is deprecated/);
     }
 
+    ['@test Ember.NAME_KEY is deprecated']() {
+      expectDeprecation(() => {
+        Ember.NAME_KEY;
+      }, 'Using `Ember.NAME_KEY` is deprecated, override `.toString` instead');
```
Comment (krisselden) [inline packages/ember/tests/reexports_test.js]: should make it clear that it is static toString.   MyClass.toString = () => 'MyClass'

File: packages/ember/index.js [hunk 4]
```diff
@@ -202,7 +202,20 @@ Ember.canInvoke = utils.canInvoke;
 Ember.tryInvoke = utils.tryInvoke;
 Ember.wrap = utils.wrap;
 Ember.uuid = utils.uuid;
-Ember.NAME_KEY = utils.NAME_KEY;
+Ember.setName = utils.setName;
```
Comment (krisselden) [inline packages/ember/index.js]: if we prefer toString on the constructor, why are we exporting this?

Comment (rwjblue) [reply packages/ember/index.js]: Derp. I missed this when I last reviewed. Good catch @krisselden.

Comment (krisselden) [reply packages/ember/index.js]: remove this and it is good to go

**Review by krisselden — DISMISSED** [review]

**Review by Turbo87 — APPROVED** [review]
rebased, and adjusted the deprecation warning to stay until v3.9.0

/cc @rwjblue
